File: projects/Swift/Tests/Examples/Create.swift
//
//  Create.swift
//  

import XCTest
@testable import ChronoxorProto

class ExampleCreate: XCTestCase {
    func testCreate() {
        // Create a new account using FBE model
        let account = AccountModel()
        let modelBegin = try! account.createBegin()
        var accountBegin = try! account.model.setBegin()
        try! account.model.id.set(value: 1)
        try! account.model.name.set(value: "Test")
        try! account.model.state.set(value: State.good)
        var walletBegin = try! account.model.wallet.setBegin()
        try! account.model.wallet.currency.set(value: "USD")
        try! account.model.wallet.amount.set(value: 1000.0)
        account.model.wallet.setEnd(fbeBegin: walletBegin)
        account.model.setEnd(fbeBegin: accountBegin)
        _ = account.createEnd(fbeBegin: modelBegin)
        assert(account.verify())

        // Show the serialized FBE size
        print("FBE size: \(account.buffer.size)")

        // Access the account using the FBE model
        let access = AccountModel()
        access.attach(buffer: account.buffer)
        assert(access.verify())

        let id: Int32
        let name: String
        let state: State
        let walletCurrency: String
        let walletAmount: Double

        accountBegin = access.model.getBegin()
        id = access.model.id.get()
        name = access.model.name.get()
        state = access.model.state.get()
        walletBegin = access.model.wallet.getBegin()
        walletCurrency = access.model.wallet.currency.get()
        walletAmount = access.model.wallet.amount.get()
        access.model.wallet.getEnd(fbeBegin: walletBegin)
        access.model.getEnd(fbeBegin: accountBegin)

        // Show account content
        print("account.id = \(id)")
        print("account.name = \(name)")
        print("account.state = \(state)")
        print("account.wallet.currency = \(walletCurrency)")
        print("account.wallet.amount = \(walletAmount)")
    }
}

File: examples/create.cpp
/*!
    \file create.cpp
    \brief Fast Binary Encoding create & access example
    \author Ivan Shynkarenka
    \date 08.05.2018
    \copyright MIT License
*/

#include "../proto/proto_models.h"

#include <iostream>

int main(int argc, char** argv)
{
    // Create a new account using FBE model
    FBE::proto::AccountModel account;
    size_t model_begin = account.create_begin();
    size_t account_begin = account.model.set_begin();
    account.model.id.set(1);
    account.model.name.set("Test");
    account.model.state.set(proto::State::good);
    size_t wallet_begin = account.model.wallet.set_begin();
    account.model.wallet.currency.set("USD");
    account.model.wallet.amount.set(1000.0);
    account.model.wallet.set_end(wallet_begin);
    account.model.set_end(account_begin);
    account.create_end(model_begin);
    assert(account.verify());

    // Show the serialized FBE size
    std::cout << "FBE size: " << account.buffer().size() << std::endl;

    // Access the account using the FBE model
    FBE::proto::AccountModel access;
    access.attach(account.buffer());
    assert(access.verify());

    int32_t id;
    std::string name;
    proto::State state;
    std::string wallet_currency;
    double wallet_amount;

    account_begin = access.model.get_begin();
    access.model.id.get(id);
    access.model.name.get(name);
    access.model.state.get(state);
    wallet_begin = access.model.wallet.get_begin();
    access.model.wallet.currency.get(wallet_currency);
    access.model.wallet.amount.get(wallet_amount);
    access.model.wallet.get_end(wallet_begin);
    access.model.get_end(account_begin);

    // Show account content
    std::cout << std::endl;
    std::cout << "account.id = " << id << std::endl;
    std::cout << "account.name = " << name << std::endl;
    std::cout << "account.state = " << state << std::endl;
    std::cout << "account.wallet.currency = " << wallet_currency << std::endl;
    std::cout << "account.wallet.amount = " << wallet_amount << std::endl;

    return 0;
}
